src/platform_impl/macos/app_delegate.rs

```diff
@@ -79,3 +106,45 @@ extern "C" fn did_resign_active(this: &Object, _: Sel, _: id) {
     }
     trace!("Completed `applicationDidResignActive`");
 }
+
+extern "C" fn application_open_file(_this: &Object, _: Sel, _sender: id, filename: id) -> BOOL {
+    trace!("Triggered `application:openFile:`");
+    let mut filenames_vec = Vec::with_capacity(1);
+    let string = ns_string_to_rust(filename);
+    filenames_vec.push(string.into());
+    let event = Event::OpenFiles(filenames_vec);
+    AppState::queue_event(EventWrapper::StaticEvent(event));
+    trace!("Completed `application:openFile:`");
+
+    // Return true to indicate to the OS that the file type is supported.
+    // (If the filetype turns out not to be supported, it's the application's
+    // responsibility to inform the user)
+    YES
+}
+
+extern "C" fn application_open_files(_this: &Object, _: Sel, _sender: id, filenames: id) {
+    use cocoa::foundation::NSArray;
+
+    #[allow(non_upper_case_globals)]
+    const NSApplicationDelegateReplySuccess: i32 = 0;
+
+    trace!("Triggered `application:openFiles:`");
+    let filenames_len = unsafe { filenames.count() };
+    let mut filenames_vec = Vec::with_capacity(filenames_len as usize);
+    for i in 0..filenames_len {
+        let filename = unsafe { filenames.objectAtIndex(i) };
+        let name_string = ns_string_to_rust(filename);
+        filenames_vec.push(name_string.into());
+    }
+    let event = Event::OpenFiles(filenames_vec);
+    AppState::queue_event(EventWrapper::StaticEvent(event));
+
+    let cls = objc::runtime::Class::get("NSApplication").unwrap();
+    let app: cocoa::base::id = unsafe { msg_send![cls, sharedApplication] };
+    // Indicate to the OS that the file types are supported.
+    // (If a filetype turns out not to be supported, it's the application's
+    // responsibility to inform the user)
+    unsafe { msg_send![app, replyToOpenOrPrint: NSApplicationDelegateReplySuccess] }
```

Comment:
What happens if we don't do that here. Do you know why this doesn't just return a boolean? That seems a bit inconsistent.

The documentation says that delegates "should" invoke this, not that they must.

Reply:
The reason why I originally put this here is that I thought that omitting this would make the OS display a message saying that the application cannot open files in the "..." format. As shown on the image [here](https://stackoverflow.com/questions/37018964/on-mac-os-x-opening-my-document-by-double-clicking-at-the-document-fails-if-my-a). But it seems that as long as the `openFile` and `openFiles` functions are implemented this message isn't displayed even if those functions indicate failure.

Although I think that the two functions as defined by winit should have identical behavior so if `openFile` returns `YES` then `openFiles` should also indicate success. The reasoning is that even if this is not what the user expects it's probably better if the behavior is consistent as opposed to "it sometimes works but sometimes it doesn't".

Reply:
If macOS assumes success by default, the behavior is identical though. That's why I'm asking if it makes any difference. If it doesn't there's no difference between the two.

Reply:
I went through the docs again in hopes of finding the answer to what's the effect of indicating success or failure. But I did not find the answer to that.

I then did some tests with a minimal Objective-C application selecting one or two files on my desktop and using the "right-click" context menu to open them with the test application. First of all, only `openFiles` was called even if I had a single file selected. I tried not calling the `replyToOpenOrPrint` function, and I tried all combinations of calling it on the `sharedApplication` and the `sender` with both success and failure. I could not see any difference on the GUI or within Console. In other words I couldn't see any difference at all.

I still believe that whatever we choose `openFile` to do, the same behavior should apply to `openFiles`.

What do you reckon, what should these functions indicate?

Reply:
Thanks for looking into this, that seems like really useful information. It sounds like it is mostly a recommendation then, based on the assumption that the application is not a toolkit and instead handles everything by itself.

> I still believe that whatever we choose openFile to do, the same behavior should apply to openFiles.

You can't return a boolean though, so that simply isn't an option. Trying to draw any parallels between the two like returning a boolean is the same as executing some callback is pointless. One does not stand in any relation to the other really.

Sure, we shouldn't return FALSE from one and return the callback with a success state, but if the callback is by default true, then returning TRUE should be fine.

Based on what I've seen, it seems like it *should* work fine to not execute that callback, however this might require some amount of trust that things aren't going to change in the future. I'm not sure that I trust in that. So the sensible and robust solution is probably to do what you're doing right now and keep it.

Reply:
Awesome possum